Framework/Algorithms/src/CalculateCountRate.cpp

```diff
@@ -0,0 +1,719 @@
+ * algorithm properties
+ *
+ *@param InputWorkspace -- event workspace to process. Also retrieves algorithm
+ *                         properties, relevant to the workspace.
+ *@return -- the input workspace cropped according to XMin-XMax ranges in units,
+ *           requested by the user
+ *
+*/
+void CalculateCountRate::setSourceWSandXRanges(
+    DataObjects::EventWorkspace_sptr &InputWorkspace) {
+
+  std::string RangeUnits = getProperty("RangeUnits");
+  auto axis = InputWorkspace->getAxis(0);
+  const auto unit = axis->unit();
+
+  API::MatrixWorkspace_sptr wst;
+  if (unit->unitID() != RangeUnits) {
+    auto conv = createChildAlgorithm("ConvertUnits", 0, 1);
+    std::string wsName = InputWorkspace->name();
+    if (wsName.empty()) {
+      wsName = "_CountRate_UnitsConverted";
+    } else {
+      wsName = "_" + wsName + "_converted";
+    }
+
+    conv->setProperty("InputWorkspace", InputWorkspace);
+    conv->setPropertyValue("OutputWorkspace", wsName);
+    std::string Emode = getProperty("Emode");
+    conv->setProperty("Emode", Emode);
+    conv->setProperty("Target", RangeUnits);
+
+    conv->setRethrows(true);
+    conv->execute();
+    wst = conv->getProperty("OutputWorkspace");
+
+  } else {
+    wst = InputWorkspace;
+  }
+  m_workingWS = boost::dynamic_pointer_cast<DataObjects::EventWorkspace>(wst);
+  if (!m_workingWS) {
+    throw std::runtime_error("SetWSDataRanges:Can not retrieve EventWorkspace "
+                             "after converting units");
+  }
+  // data ranges
+  m_XRangeMin = getProperty("XMin");
+  m_XRangeMax = getProperty("XMax");
+
+  if (m_XRangeMin == EMPTY_DBL() && m_XRangeMax == EMPTY_DBL()) {
+    m_rangeExplicit = false;
+  } else {
+    m_rangeExplicit = true;
+  }
+
+  double realMin, realMax;
+  m_workingWS->getEventXMinMax(realMin, realMax);
+  if (!m_rangeExplicit) { // The range is the whole workspace range
+    m_XRangeMin = realMin;
+    // include rightmost limit into the histogramming
+    m_XRangeMax = realMax * (1. + std::numeric_limits<double>::epsilon());
+    return;
+  }
+
+  if (m_XRangeMin == EMPTY_DBL()) {
+    m_XRangeMin = realMin;
+  }
+  if (m_XRangeMax == EMPTY_DBL()) {
+    // include rightmost limit into the histogramming
+    m_XRangeMax = realMax * (1. + std::numeric_limits<double>::epsilon());
+  }
+  if (m_XRangeMin < realMin) {
+    g_log.debug() << "Workspace constrain min range changed from: "
+                  << m_XRangeMin << " To: " << realMin << std::endl;
+    m_XRangeMin = realMin;
+  }
+  if (m_XRangeMax > realMax) {
+    g_log.debug() << "Workspace constrain max range changed from: "
+                  << m_XRangeMax << " To: " << realMax << std::endl;
+    m_XRangeMax = realMax * (1. + std::numeric_limits<double>::epsilon());
+  }
+  // check final ranges valid
+  if (m_XRangeMin > m_XRangeMax) {
+    throw std::invalid_argument(
+        " Minimal spurion search data range is bigger than the maximal range.");
```

Comment:
It's good that these tests are now here, but when I give only the XMax value, and it happens to be too small, the error I see is this 'Minimal spurion search data range is bigger than the maximal range.' which may be a bit confusing. I would expect the error message to be more like 'XMax too small'. Would it make sense to improve the error reporting in this case?